File: system-test/dotnet/SystemTests/TestInfrastructure/Configuration/PropertyLoader.cs
using Dsl.Port;
using Dsl.Core.Shared;

namespace SystemTests.TestInfrastructure.Configuration;

public static class PropertyLoader
{
    public static Environment GetEnvironment(Environment? fixedEnvironment)
    {
        if (fixedEnvironment != null)
        {
            return fixedEnvironment.Value;
        }

        var environmentMode = GetRequiredEnvironmentVariable("ENVIRONMENT", "local|acceptance|qa|production");
        return Enum.Parse<Environment>(environmentMode, ignoreCase: true);
    }

    public static ExternalSystemMode GetExternalSystemMode(ExternalSystemMode? fixedExternalSystemMode)
    {
        if (fixedExternalSystemMode != null)
        {
            return fixedExternalSystemMode.Value;
        }

        var externalSystemMode = GetRequiredEnvironmentVariable("EXTERNAL_SYSTEM_MODE", "stub|real");
        return Enum.Parse<ExternalSystemMode>(externalSystemMode, ignoreCase: true);
    }

    private static string GetRequiredEnvironmentVariable(string variableName, string allowedValues)
    {
        var value = System.Environment.GetEnvironmentVariable(variableName);

        if (string.IsNullOrEmpty(value))
        {
            throw new InvalidOperationException(
                $"Environment variable '{variableName}' is not defined. Please set {variableName}=<{allowedValues}>");
        }

        return value;
    }
}

File: system-test/java/src/main/java/com/optivem/shop/systemtest/configuration/PropertyLoader.java
package com.optivem.shop.systemtest.configuration;

import com.optivem.shop.dsl.port.ChannelMode;

public class PropertyLoader {
    private PropertyLoader() {
    }

    public static Environment getEnvironment(Environment fixedEnvironment) {
        if (fixedEnvironment != null) {
            return fixedEnvironment;
        }

        var environmentMode = getRequiredSystemProperty("environment", "local|acceptance|qa|production");
        return Environment.valueOf(environmentMode.toUpperCase());
    }

    public static ExternalSystemMode getExternalSystemMode(ExternalSystemMode fixedExternalSystemMode) {
        if (fixedExternalSystemMode != null) {
            return fixedExternalSystemMode;
        }

        var externalSystemMode = getRequiredSystemProperty("externalSystemMode", "stub|real");
        return ExternalSystemMode.valueOf(externalSystemMode.toUpperCase());
    }

    public static ChannelMode getChannelMode(ChannelMode fixedChannelMode) {
        if (fixedChannelMode != null) {
            return fixedChannelMode;
        }

        var value = System.getProperty("channelMode");
        if (value == null || value.isBlank()) {
            return ChannelMode.DYNAMIC;
        }
        return ChannelMode.valueOf(value.toUpperCase());
    }

    private static String getRequiredSystemProperty(String propertyName, String allowedValues) {
        var value = System.getProperty(propertyName);
        if (value == null || value.isBlank()) {
            throw new IllegalStateException(
                    String.format("System property '%s' is not defined. Please specify -D%s=<%s>",
                            propertyName, propertyName, allowedValues)
            );
        }
        return value;
    }
}
